File: app/src/main/java/tk/lorddarthart/rxnewstestapp/HttpServiceHelper.java
package tk.lorddarthart.rxnewstestapp;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

class HttpServiceHelper {
    private static HttpServiceHelper mInstance;
    private static final String BASE_URL = "https://drive.google.com/";
    private Retrofit mRetrofit;

    private HttpServiceHelper() {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient.Builder client = new OkHttpClient.Builder()
                .addInterceptor(interceptor);

        mRetrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
    }

    static HttpServiceHelper getInstance() {
        if (mInstance == null) {
            mInstance = new HttpServiceHelper();
        }
        return mInstance;
    }

    JSONPlaceHolderApi getJSONApi() {
        return mRetrofit.create(JSONPlaceHolderApi.class);
    }
}

File: app/src/main/java/tk/lorddarthart/rxnewstestapp/util/network/HttpServiceHelper.kt
package tk.lorddarthart.rxnewstestapp.util.network

import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import tk.lorddarthart.rxnewstestapp.util.constants.Constants.BASE_URL
import tk.lorddarthart.rxnewstestapp.util.network.retrofit.GetNewsList

internal class HttpServiceHelper private constructor() {
    private val retrofit: Retrofit
    private val client: OkHttpClient

    private val interceptor: HttpLoggingInterceptor = HttpLoggingInterceptor()

    val newsApi: GetNewsList
        get() = retrofit.create(GetNewsList::class.java)

    init {
        interceptor.level = HttpLoggingInterceptor.Level.BODY

        client = OkHttpClient.Builder()
                .addInterceptor(interceptor).build()

        retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapter.create())
                .client(client)
                .build()
    }

    companion object {
        private val TAG = HttpServiceHelper::class.java.simpleName

        lateinit var instance: HttpServiceHelper
    }
}
